code/site/libraries/authentication.php

```diff
@@ -273,4 +273,118 @@ private static function getAuthorizationHeader()
 
 		return $headers;
 	}
+
+	/**
+	 * Find if the user is trying imporsonate other user
+	 *
+	 * @return  int|string|NULL  User id or Email id or null
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public static function getImpersonateHeader()
+	{
+		$jinput           = JFactory::getApplication()->input;
+		$xImpersonate     = $jinput->server->get('X-Impersonate', '', 'STRING');
+		$httpXImpersonate = $jinput->server->get('HTTP_X_IMPERSONATE', '', 'STRING');
+
+		if (!empty($xImpersonate))
+		{
+			return $xImpersonate;
+		}
+		elseif (!empty($httpXImpersonate))
+		{
+			return $httpXImpersonate;
+		}
+	}
+
+	/**
+	 * Find if the user is trying imporsonate other user
+	 *
+	 * @param   int  $tokenUserId  The userid for which token hash is validated
+	 *
+	 * @return  int|boolean  User id or Email id or null
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public static function getUserIdToImpersonate($tokenUserId)
+	{
+		// Get user for this key
+		$user         = JFactory::getUser($tokenUserId);
+		$isSuperAdmin = $user->authorise('core.admin');
+
+		// If this user is not super admin user, return false
+		if (!$isSuperAdmin)
+		{
+			return false;
+		}
+
+		// User is superadmin, lets find out if he/she is trying to imporsonate other user
+		$userToImpersonate = self::getImpersonateHeader();
+
+		// If other is to be impersonated
+		if ($userToImpersonate)
```

Comment:
```suggestion
		if (!$userToImpersonate) { return false; }
```

followed by the code currently inside the if condition.